src/silo/preprocessing/preprocessor.cpp

```diff
@@ -0,0 +1,515 @@
+            )
+         );
+
+         auto return_code = preprocessing_db.query(fmt::format(
+            R"-(
+create or replace view nuc_{0} as
+select key, sequence,
+partitioned_metadata.partition_id as partition_id
+{2}
+from raw_nuc_{0} raw, partitioned_metadata
+where raw.key = partitioned_metadata.{1};)-",
+            seq_name,
+            database_config.schema.primary_key,
+            order_by_select
+         ));
+
+         if (return_code->HasError()) {
+            SPDLOG_ERROR(return_code->GetError());
+            throw PreprocessingException(return_code->GetError());
+         }
+      }
+
+      for (const auto& [seq_name, reference_sequence] : reference_genomes.raw_aa_sequences) {
+         preprocessing_db.generateNucSequenceTable(
+            "raw_gene_" + seq_name,
+            reference_sequence,
+            preprocessing_config.getGeneFilenameNoExtension(seq_name).replace_extension(
+               silo::preprocessing::FASTA_EXTENSION
+            )
+         );
+         silo::FastaReader fasta_reader(preprocessing_config.getGeneFilenameNoExtension(seq_name)
+                                           .replace_extension(silo::preprocessing::FASTA_EXTENSION)
+         );
+
+         auto return_code = preprocessing_db.query(fmt::format(
+
+            R"-(
+create or replace view gene_{0} as
+select key, sequence,
+partitioned_metadata.partition_id as partition_id
+{2}
+from raw_gene_{0} raw, partitioned_metadata
+where raw.key = partitioned_metadata.{1};
+)-",
+            seq_name,
+            database_config.schema.primary_key,
+            order_by_select
+         ));
+      }
+   }
+
+   preprocessing::Partitions partition_descriptor = preprocessing_db.getPartitionDescriptor();
+
+   SPDLOG_INFO("preprocessing - building database");
+
+   return buildDatabase(partition_descriptor, reference_genomes, order_by_clause, alias_key);
+}
+
+void Preprocessor::buildPartitioningTable() {
+   if (database_config.schema.partition_by.has_value()) {
+      SPDLOG_INFO("preprocessing - calculating partitions");
+
+      auto return_code = preprocessing_db.query(fmt::format(
+         R"-(
+create
+or replace table partition_keys as
+select row_number() over () - 1 as id, partition_key, count
+from (SELECT {} as partition_key, COUNT(*) as count
+      FROM metadata_table
+      GROUP BY partition_key
+      ORDER BY partition_key);
+)-",
+         database_config.schema.partition_by.value()
+      ));
+
+      if (return_code->HasError()) {
+         throw PreprocessingException(
+            "Error in the execution of the duckdb statement for partition key table "
+            "generation: " +
+            return_code->GetError()
+         );
+      } else {
+         SPDLOG_TRACE("Executed statement for partition key table generation.");
+         SPDLOG_TRACE(return_code->ToString());
+      }
+
+      return_code = preprocessing_db.query(
+         R"-(
+create or replace table partitioning as
+with recursive
```

Comment:
What does `with recursive` do?

Reply:
It is part of the SQL I will write comments by function names as you suggested